iothub/device/samples/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -181,79 +181,47 @@ private async Task OnC2dMessageReceivedAsync(Message message, object context)
         // has a void return type. As a result, any operation within this block will be executed unmonitored on another thread.
         // To prevent multi-threaded synchronization issues, the async method InitializeClientAsync being called in here first grabs a lock before attempting to
         // initialize or dispose the device client instance; the async method GetTwinAndDetectChangesAsync is implemented similarly for the same purpose.
-        private async void ConnectionStatusChangeHandler(ConnectionInfo connectionInfo)
+        private async void ConnectionStatusChangeHandlerAsync(ConnectionInfo connectionInfo)
         {
-            var status = connectionInfo.Status;
-            var reason = connectionInfo.ChangeReason;
-            Console.WriteLine($"Connection status changed: status={status}, reason={reason}");
+            ConnectionStatus status = connectionInfo.Status;
+            ConnectionStatusChangeReason reason = connectionInfo.ChangeReason;
+            Console.WriteLine($"Connection status changed: status={status}, reason={reason}, recommendation={connectionInfo.RecommendedAction}");
+
+            // In our case, we can operate with more than 1 shared access key and attempt to fall back to a secondary.
+            // We'll disregard the SDK's recommendation and attempt to connect with the second one.
+            if (status == ConnectionStatus.Disconnected
```

Comment:
Because this is an exception to using the recommended action.

Putting it inside recommended action means if the mapping was incorrect, the code would never run.

I prefer it up top for that reason.